### PR TITLE
fix(typing): update to latest version of Pyright and fix errors

### PR DESCRIPTION
**Pull request by benhoyt**
So annoying how quickly this tooling changes. Couple of notes:

- ops/charm.py: it seems like the new version disallows overriding the type in a subclass, so just use a "pyright: ignore" comment to work around this. Maybe there's a better way.
- The new Pyright no longer allows a type variable to be defined inside a function. Not sure what the rationale for this was, but they have to be moved to the top level (away from where they're used!) now.
- Some of the bytes/str/bytearray stuff is annoying. Maybe there are better solutions for these types. Oh for Go's io.Reader.

Suggestions to improve any of these welcome.

### CODE REVIEW

**Review by tonyandrewmeyer — COMMENTED** [review]
A couple of small suggestions and a couple of small questions, but overall looks good to me.

File: pyproject.toml [hunk 1]
```diff
@@ -38,4 +38,5 @@ reportMissingModuleSource = false
 reportPrivateUsage = false
 reportUnnecessaryIsInstance = false
 reportUnnecessaryComparison = false
+disableBytesTypePromotions = false
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply pyproject.toml:41]: That's odd. When I comment out that line and leave it as the default, Pyright prints a bunch of this message:

```
$ tox -e static
static: commands[0]> pyright
/home/ben/w/operator/ops/pebble.py
  /home/ben/w/operator/ops/pebble.py:1617:21 - error: Unnecessary "cast" call; type is already "bytes" (reportUnnecessaryCast)
  /home/ben/w/operator/ops/pebble.py:2993:52 - error: Argument of type "bytearray" cannot be assigned to parameter "buf" of type "bytes" in function "_next_part_boundary"
    "bytearray" is incompatible with "bytes"
...
```

Oh wait, it looks like the default is actually true in "strict" mode: https://github.com/microsoft/pyright/blob/main/docs/configuration.md#diagnostic-rule-defaults

Makes me wonder if we're causing ourselves a bunch of extra pain by being in strict mode...

File: ops/charm.py [hunk 2]
```diff
@@ -466,7 +466,7 @@ class RelationCreatedEvent(RelationEvent):
     can occur before units for those applications have started. All existing
     relations should be established before start.
     """
-    unit: None
+    unit: None  # pyright: ignore[reportIncompatibleVariableOverride]
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply ops/charm.py:469]: Hmmm, I had a play with this, and I don't like the fact that then the docs for `RelationEvent` wouldn't have "unit" defined at all, which seems very weird. Or were you suggesting something else?

File: test/test_charm.py [hunk 3]
```diff
@@ -101,12 +101,11 @@ def test_observe_decorated_method(self):
...
         events: typing.List[ops.EventBase] = []
 
-        def dec(fn: typing.Callable[['MyCharm', ops.EventBase], None]  # noqa: F821
-                ) -> typing.Callable[..., None]:
+        def dec(fn: typing.Any) -> typing.Any:
             # simple decorator that appends to the nonlocal
             # `events` list all events it receives
             @functools.wraps(fn)
-            def wrapper(charm: 'MyCharm', evt: ops.EventBase):
+            def wrapper(charm: typing.Any, evt: typing.Any):
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply test/test_charm.py:108]: Good call -- updated with the more minimal change.

File: test/test_model.py [hunk 4]
```diff
@@ -2381,6 +2381,13 @@ def test_unresolved_ingress_addresses(self):
 _metric_and_label_pair = typing.Tuple[typing.Dict[str, float], typing.Dict[str, str]]
 
 
+ValidMetricsTestCase = typing.Tuple[
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply test/test_model.py:2384]: No, fair call -- updated both.

File: ops/charm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tonyandrewmeyer) [reply ops/charm.py:469]: No, that was what I was suggesting, and I indeed wondered if that meant it was not possible to have it in the `RelationEvent` doc (unless there's some Sphinx trick for this I don't know). I agree that's not great.

In balance, I'm fine keeping it how you have it with the ignores. The goal was to have the types correct in the various subclasses so people didn't need a bunch of `assert unit is not None` and that all still works, so there doesn't seem to be much value in trying too hard to get rids of the ignores.

**Review by tonyandrewmeyer — APPROVED** [review]
Interesting thought about strict! Maybe something to look at next time we need to do this :smile:.

For now, looks great to me!

**Comment by benhoyt** [comment]
Updates for new version of Pyright, no user-facing changes, merged without further review.

**Pull request merged**